lib/similarity_utils.py

```diff
@@ -21,3 +22,70 @@ def iou(lista1: List[Any], lista2: List[Any]) -> float:
         intersection = set(lista1).intersection(set(lista2))
         union = set(lista1).union(set(lista2))
         return len(intersection) / len(union)
+
+
+def sort_two_lists(list1: List[Any], list2: List[Any], reverse: bool = True):
+    """metoda sortuje dwie listy równocześnie
+    list1 jest sortowana po wartościach, a kolejność elementów w list2 zależy od sortowania list1
+    reverse = True oznacza kolejność malejącą, a reverse = False oznacza kolejność rosnącą"""
+    zipped_lists = zip(list1, list2)
+    list1, list2 = zip(*sorted(zipped_lists, reverse=reverse))
+    return list1, list2
+
+
+def find_actor(data: pd.DataFrame, actor_id: str) -> List[Any]:
+    """metoda na podstawie ramki danych oraz id aktora, zwraca jego dane z ramki danych"""
+    actor_index = data[data['nconst'] == actor_id].index
+    return data.iloc[actor_index[0]]
+
+
+def prepare_pandas_row(pandas_row: pd.DataFrame) -> List[Any]:
+    """metoda zamienia wiersz wydobyty z Pandas DataFrame i przekształca go w pythonową listę"""
+    p_list = []
+    for value in pandas_row:
+        try:
+            p_list.append(value.tolist())
+        except AttributeError:
+            p_list.append(value)
+    return p_list
+
+
+def similarity(actor1: List[Any], actor2: List[Any]) -> float:
+    """metoda licząca similarity pomiędzy dwoma aktorami
+    jej argumentami są dwie listy, a wartością wyjściową wartość z przedziału [-1, 1]
+    metoda jest przygotowana pod dane ze zbioru JOINED_DATA"""
+    weights = [0.3, 0.2, 0.3, 0.2]
+    values = [
+        iou(actor1[1], actor2[1]),           # similarty ze względu na ilość wspólnych filmów
+        iou(actor1[2], actor2[2]),           # similarity ze względu na rodzaj granych produkcji
+        iou(actor1[7], actor2[7]),           # similarity ze względu na gatunek granych produkcji
+        1 if actor1[8] == actor2[8] else 0   # similarity ze względu na tę samą płeć
+    ]
+    #TODO dodać linijkę uwzględniającą kolumnę knownForTitles za pomocą metody iou
+    return sum(weights[i] * values[i] for i in range(4)) * 2 - 1
```

Comment:
Tutaj wartość $4$ wystepuje w `range(4)` ponieważ tyle mam kolumn od których zależy similarity. Czy optymalniej, ze strony kodu, jest przy każdej modyfikacji i dodaniu nowej kolumny oraz wagi zwiększać z palca tę liczbę w kodzie? Jak na moje, to jej wyliczanie jako `len(weights)` albo `len(values)` jest po prostu niepotrzebne. Zostawić jak jest?

Reply:
jest - inaczej 4 działa jak antywzorzec magicznej liczby. Sugeruje ponadto assercję na to by weights in values miało równy rozmiar.

Reply:
Dodano asercję. Nie do końca rozumiem resztę komentarza

Reply:
Magiczna liczba to antywzorzec programowania. Chodzi o stałą wartość, która nie posiada czytelnej interpretacji dla dowolnej osoby czytającej dany kod. https://en.wikipedia.org/wiki/Magic_number_(programming) 

Tak zadana 4 jest trudna do interpretacji jako `len(values)` ma czytelny sens.

Reply:
Zmienione na $4$ na  `len(weights)`